frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,230 @@
+                text = T(_("The user dictionary\n%1\nis not alphabetically sorted.\n\nIt has been disabled."), name),
+            })
+            logger.warn("UserHyph: Dictionary " .. name .. " is not sorted alphabetically.")
+            G_reader_settings:makeFalse("hyph_user_dict")
+        elseif ret == self.USER_DICT_MALFORMED then
+            UIManager:show(InfoMessage:new{
+                text = T(_("The user dictionary\n%1\nhas corrupted entries.\n\nOnly valid entries will be used."), name),
+            })
+            logger.warn("UserHyph: Dictionary " .. name .. " has corrupted entries.")
+        end
+    else
+        self.ui.document:setUserHyphenationDict() -- clear crengine user hyph dict
+    end
+end
+
+-- Reload on change of the hyphenation language
+function ReaderUserHyph:onTypographyLanguageChanged()
+    self:loadUserDictionary()
+end
+
+-- Reload on "ChangedUserDictionary" event,
+-- doesn't load dictionary if filesize and filename haven't changed
+-- if reload==true reload
+function ReaderUserHyph:loadUserDictionary(reload)
+    self:loadDictionary(self:isAvailable() and self:getDictionaryPath() or "", reload and true or false)
+    self.ui:handleEvent(Event:new("UpdatePos"))
+end
+
+-- Functions to use with the UI
+
+function ReaderUserHyph:isAvailable()
+    return G_reader_settings:isTrue("hyph_user_dict") and self:_enabled()
+end
+
+function ReaderUserHyph:_enabled()
+    return self.ui.typography.hyphenation
+end
+
+-- add Menu entry
+function ReaderUserHyph:getMenuEntry()
+    return {
+        text = _("Custom hyphenation rules"),
+        help_text = _("The hyphenation of a word can be changed from its default by long pressing for 3 seconds and selecting 'Hyphenate'."),
+        callback = function()
+            local hyph_user_dict =  not G_reader_settings:isTrue("hyph_user_dict")
+            G_reader_settings:saveSetting("hyph_user_dict", hyph_user_dict)
+            self:loadUserDictionary() -- not needed to force a reload here
+        end,
+        checked_func = function()
+            return self:isAvailable()
+        end,
+        enabled_func = function()
+            return self:_enabled()
+        end,
+        separator = true,
+    }
+end
+
+-- Helper functions for dictionary entries-------------------------------------------
+
+-- checks if suggestion is well formated
+function ReaderUserHyph:checkHyphenation(suggestion, word)
+    if suggestion:find("%-%-") then
+        return false -- two or more consecutive '-'
+    end
+
+    suggestion = suggestion:gsub("-","")
+    if self.ui.document:getLowercasedWord(suggestion) == self.ui.document:getLowercasedWord(word) then
+        return true -- characters match (case insensitive)
+    end
+    return false
+end
+
+function ReaderUserHyph:updateDictionary(word, hyphenation)
+    local dict_file = self:getDictionaryPath()
+    local new_dict_file = dict_file .. ".new"
+    local dict
+    local new_dict
```

Comment:
No need now to declare them local way that early, do it when creating the variables.